File: java/src/main/java/it/giannibombelli/wsc2026/payment/domain/events/TransactionRejected.java
package it.giannibombelli.wsc2026.payment.domain.events;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import it.giannibombelli.wsc2026.payment.domain.payment.Provider;
import it.giannibombelli.wsc2026.payment.domain.payment.TransactionId;

public record TransactionRejected(
    PaymentId aggregateId,
    Provider provider,
    TransactionId transactionId,
    Description reason
) implements PaymentEvent {
    public TransactionRejected {
        Require.requireArgument(aggregateId, "aggregateId");
        Require.requireArgument(provider, "provider");
        Require.requireArgument(transactionId, "transactionId");
        Require.requireArgument(reason, "reason");
    }
}

File: typescript/src/payment/domain/events/transactionRejected.ts
// Evento emesso quando una transazione viene rifiutata dal provider.

import type { Event } from '@/common/domain/model/event.js';
import type { Description } from '@/common/domain/primitive/description.js';
import type { PaymentId } from '../payment/paymentId.js';
import type { Provider } from '../payment/provider.js';
import type { TransactionId } from '../payment/transactionId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type TransactionRejected = Event<'TransactionRejected', PaymentId> & {
  readonly provider: Provider;
  readonly transactionId: TransactionId;
  readonly reason: Description;
};

export function transactionRejected(
  aggregateId: PaymentId,
  provider: Provider,
  transactionId: TransactionId,
  reason: Description
): TransactionRejected {
  requireArgument(aggregateId, 'aggregateId');
  requireArgument(provider, 'provider');
  requireArgument(transactionId, 'transactionId');
  requireArgument(reason, 'reason');
  return { kind: 'TransactionRejected', aggregateId, provider, transactionId, reason };
}
